go/workflow-interceptor-updates.md

```diff
@@ -0,0 +1,417 @@
+Current:
+
+* A concept called a context propagator handles marshalling to/from workflow and activity proto "headers" (a construct
+  in the body not to be confused with actual HTTP headers)
+* Client options accept an OpenTracing tracer
+* An internal context propagator is creating for the OpenTracing tracer that marshals to/from a span context to a
+  special key in the headers
+* The tracer is sent all throughout the contextual work of workflows and activities so that spans can be created at
+  different times
+
+Proposal:
+
+* A new ability to retrieve a mutable set of proto "headers" on the context will be made so that interceptors can set
+  workflow/activity headers in either direction
+  * Note, the reason this is on the context instead of the input is because it is contextual information that flows
+    through the system instead of literal parameters. We do the same for other bits of contextual information.
+* Internally, headers will be obtained from the context by the last interceptor
+  * So headers are set on the context before interceptor invocation for mutation by interceptors
+* A `tracing` package will be created with:
+  * `WorkerInterceptor` implementations for OpenTracing and OpenTelemetry tracers
+    * Wraps the same inbound/outbound calls in spans that are otherwise wrapped by the tracer today
+    * On inbound `ExecuteActivity` and `ExecuteWorkflow` span context is unmarshaled from header
+    * On outbound `ExecuteActivity`, `ExecuteLocalActivity`, `ExecuteChildWorkflow`, and `NewContinueAsNewError` span
+      context is marshaled into header
+  * `Client` implementations for OpenTracing and OpenTelemetry tracers
+    * Wraps the same calls in spans that are otherwise wrapped by the tracer today
+    * On `ExecuteWorkflow` and `SignalWithStartWorkflow` span context is marshaled into header
+
+```go
+package interceptors
+
+// Header returns nil if this context does not have the header on it (meaning
+// they cannot be mutated by an interceptor).
+func Header(ctx interface { Value(interface{}) interface{} }) map[string]*commonpb.Payload
+```
+
+```go
+package tracing
+
+func NewOpenTracingClient(c client.Client, tracer opentracing.Tracer) client.Client
+func NewOpenTracingInterceptor(tracer opentracing.Tracer) interceptors.WorkerInterceptor
+
+func NewOpenTelemetryClient(c client.Client, tracer trace.Tracer) client.Client
+func NewOpenTelemetryInterceptor(tracer trace.Tracer) interceptors.WorkerInterceptor
+```
+
+For discussion:
+
+* Is the approach of `workflow.Header` and `activity.Header` acceptable or do we want to combine them somehow?
+* This effectively deprecates the `ClientOptions.Tracer` field, shall we remove it completely?
+  * If not, should we gut the existing `opentracing` impl and have it default to this implementation (i.e. if
+    `ClientOptions.Tracer` is present, implicitly wrap the client and add the interceptor)?
+* This effectively deprecates `ContextPropagator`s, shall we remove them completely?
```

Comment:
I believe so yes, but, I think there's a question about how much incompatible changes we're willing to make. We should get feedback from @mfateev

Reply:
What did we decide here?